etc/convert-test-results.swift

```diff
@@ -0,0 +1,332 @@
+                .replacingOccurrences(of: "&", with: "﹠")
+
+            output += 
+                """
+                <failure message="\(escapedFailure)"></failure>\n
+                """
+        }
+
+        output += "</testcase>\n"
+        return output
+    }
+}
+
+// Top-level redundant suites that we don't need to put into the xunit output.
+let ignoreSuites = [
+    // macOS
+    "AllTests",
+    "mongo-swift-driverPackageTests.xctest",
+    // linux
+    "All tests",
+    "debug.xctest"
+]
+
+/// An error thrown while parsing test output.
+struct ParsingError: LocalizedError {
+    let message: String
+
+    public var errorDescription: String? { self.message }
+
+    init(_ message: String) {
+        self.message = message
+    }
+}
+
+extension NSTextCheckingResult {
+    func readMatch(at position: Int, in line: String) throws -> String {
+        guard let range = Range(self.range(at: position), in: line) else {
+            throw ParsingError("No capture group match at position \(position)")
+        }
+        return String(line[range])
+    }
+}
+
+extension TimeInterval {
+    init(input: String) throws {
+        guard let time = TimeInterval(input) else {
+            throw ParsingError("unable to parse TimeInterval from \(input)")
+        }
+        self = time
+    }
+}
+
+func ensureSuiteMatches(old: String, new: String) throws {
+    guard old == new else {
+        throw ParsingError("test suite name \(new) does not match previously found name for current suite \(old)")
+    }
+}
+
+/// State machine which processes `swift test` output and updates itself accordingly.
+enum ParsingState {
+    /// In the following cases:
+    /// - `completeTests` is stored whenever we are in the middle of a suite, and contains any test cases we have found
+    ///   for the suite so far.
+    /// - `completeSuites` is always stored and contains all of the suites we have fully parsed so far.
+
+    /// We have read a line indicating that a test suite with the given name started, but we are not in a particular
+    /// test case.
+    case inSuite(name: String, completeTests: [TestCase], completeSuites: [TestSuite])
+    /// We've read a line indicating that a suite with the given name ended, and are expecting to read a next line
+    /// containing test pass/fail counts for the suite.
+    case awaitingSuiteDetails(name: String, completeTests: [TestCase], completeSuites: [TestSuite])
+    /// We are in the middle of a test with the given name in the given suite. `output` contains any output the test
+    /// case has produced.
+    case inTest(suite: String, name: String, output: [String], completeTests: [TestCase], completeSuites: [TestSuite])
+    /// None of the above. This means we are between test suites.
+    case none(completeSuites: [TestSuite])
+
+    // account for variation in formatting of test output on platforms. this assumes you are running this script on the
+    // same platform where you ran the tests.
```

Comment:
this is super frustrating and I think one of those warts exposed by the ObjC implementation on macOS. I remember looking into this a while back for some reason.